addons/hr_expense/models/hr_expense.py

```diff
@@ -146,24 +153,26 @@ def _compute_is_ref_editable(self):
             else:
                 expense.is_ref_editable = is_account_manager
 
-    @api.onchange('product_id', 'company_id')
-    def _onchange_product_id(self):
-        if self.product_id:
-            if not self.name:
-                self.name = self.product_id.display_name or ''
-            self.unit_amount = self.product_id.price_compute('standard_price')[self.product_id.id]
-            self.product_uom_id = self.product_id.uom_id
-            self.tax_ids = self.product_id.supplier_taxes_id.filtered(lambda tax: tax.company_id == self.company_id)  # taxes only from the same company
-            account = self.product_id.product_tmpl_id.with_company(self.company_id)._get_product_accounts()['expense']
-            if account:
-                self.account_id = account
-
-    @api.onchange('company_id')
-    def _onchange_expense_company_id(self):
-        self.employee_id = self.env['hr.employee'].search([('user_id', '=', self.env.uid), ('company_id', '=', self.company_id.id)])
+    @api.depends('product_id', 'company_id')
+    def _compute_from_product_id_company_id(self):
+        for expense in self:
+            if expense.product_id:
+                if not expense.name:
+                    expense.name = expense.product_id.display_name or ''
+                expense.unit_amount = expense.product_id.price_compute('standard_price')[expense.product_id.id]
+                expense.product_uom_id = expense.product_id.uom_id
+                expense.tax_ids = expense.product_id.supplier_taxes_id.filtered(lambda tax: tax.company_id == expense.company_id)  # taxes only from the same company
+                account = expense.product_id.product_tmpl_id.with_company(expense.company_id)._get_product_accounts()['expense']
+                if account:
+                    expense.account_id = account
+
+    @api.depends('company_id')
+    def _compute_employee_id(self):
+        for expense in self:
+            expense.employee_id = self.env['hr.employee'].search([('user_id', '=', self.env.uid), ('company_id', '=', expense.company_id.id)])
```

Comment:
`hr/models/res_users.py`:
```python
    @api.depends('employee_ids')
    @api.depends_context('company')
    def _compute_company_employee(self):
```
--> A value in cache by env.company ;).

Reply:
Not sure what's your current understanding of the ORM cache in general, but here is a quick explanation:
Every time you request a field (e.g. `partner.name`), the ORM queries the database, fetches every field of `partner` and store them in memory, in the cache. At the next `partner.name` or `partner.whatever` it won't query the database. It gets the result from memory which is obsiously much (much) faster.

One more important thing: when you iterate over a recordset then reads a fields, the ORM prefetches fields for the entire recordset.
```python
for partner in self:
    # at the first iteration, this reads every fields of all
    # records in self in one database query.
    # Following iteration reads the value from cache
    name = partner.name
```
For the computed field here `employee_id` which depends on the company, I didn't know if recomputing the field (because the company changed) overwrite the value in cache or if it was creating a new cache entry.

Reply:
@LucasLefevre Thanks for the explanations

Reply:
`Context-dependent` fields have a deeper cache, where a different value is stored w.r.t. `depends_context` field attribute ;). (See Cache implementation in api.py)

```python
a = fields.Char(compute='compute', stored=False)

@api.depends_context('company', 'other_key')
def compute(self):
  self.a = 'x' if self.env.company == 1 else 'y'

####################
record.a 
# a is computed with current record context

record.with_company(2).a
# a is recomputed with new context

record.with_company(3).a
# a is recomputed with new context

# Note that the field is only recomputed if the depends_context keys
# are different in the current context.
record.with_context(key="b").a  # no recompute, key is not in `depends_context`
record.with_context(other_key='b').a # field is recomputed
```

To go back to the `employee_id` field case:
When the context is updated by `with_company`, the field will be recomputed if a value for the given record, field and company context isn't already in cache, because the field compute method is specified as `@api.depends_context('company')`.

##### Precisions:
* `company_dependent=True` fields are `depends_context=['company']` by default.
* `translate=True` fields are `depends_context=['lang']` by default.

Reply:
@Feyensv Thanks